src/backend/distributed/utils/metadata_cache.c

```diff
@@ -2598,19 +2598,28 @@ InitializeCaches(void)
 
 	if (!performedInitialization)
 	{
-		/* set first, to avoid recursion dangers */
-		performedInitialization = true;
+		PG_TRY();
+		{
+			/* set first, to avoid recursion dangers */
+			performedInitialization = true;
+
+			/* make sure we've initialized CacheMemoryContext */
+			if (CacheMemoryContext == NULL)
+			{
+				CreateCacheMemoryContext();
+			}
 
-		/* make sure we've initialized CacheMemoryContext */
-		if (CacheMemoryContext == NULL)
+			InitializeDistTableCache();
+			RegisterForeignKeyGraphCacheCallbacks();
+			RegisterWorkerNodeCacheCallbacks();
+			RegisterLocalGroupIdCacheCallbacks();
+		}
+		PG_CATCH();
 		{
-			CreateCacheMemoryContext();
+			performedInitialization = false;
+			PG_RE_THROW();
```

Comment:
I'm not sure this PG_CATCH adds anything over moving the assignment to the end of the function such that it is only reached when we manage to complete all function calls above.

What we could use a PG_CATCH for is to reset the cache if we fail to initialise it.